File: 1448_queue.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
  int goodNodes(TreeNode* root) {
    int count = 0;
    queue<pair<TreeNode*, int>> pairs;
    pairs.push({root, root->val});
    
    while (!pairs.empty()) {
      auto next = pairs.front();
      TreeNode* node = next.first;
      int max_val = next.second;
      
      if (node->left == nullptr && node->right == nullptr) {
        pairs.pop();
        if (node->val >= max_val)
          count++;
      } else if (node->left != nullptr) {
        pairs.push({node->left, max(max_val, node->val)});
        node->left = nullptr;
      } else if (node->right != nullptr) {
        pairs.push({node->right, max(max_val, node->val)});
        node->right = nullptr;
      }
    }
    return count;
  }
};

File: 1448_queue.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
function goodNodes(root) {
  let count = 0;
  let pairs = [];
  pairs.push([root, root.val])
  
  while (pairs.length) {
    let next = pairs[pairs.length - 1];
    let node = next[0];
    let max_val = next[1];
    
    if (node.left === null && node.right === null) {
      pairs.pop();
      if (node.val >= max_val)
        count++;
    } else if (node.left !== null) {
      pairs.push([node.left, max_val >= node.val ? max_val : node.val]);
      node.left = null;
    } else if (node.right !== null) {
      pairs.push([node.right, max_val >= node.val ? max_val : node.val]);
      node.right = null;
    }
  }
  return count;
};
